### PR TITLE
Allow the creation of 1-arity tuples

### PR DESCRIPTION
**Pull request by TyOverby**
Replaces #20217

closes https://github.com/dotnet/roslyn/issues/14304

### CODE REVIEW

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 1]
```diff
@@ -920,7 +920,7 @@ public INamedTypeSymbol GetTypeByMetadataName(string fullyQualifiedMetadataName)
                 throw new ArgumentNullException(nameof(elementTypes));
             }
 
-            if (elementTypes.Length <= 1)
+            if (elementTypes.Length == 0)
             {
                 throw new ArgumentException(CodeAnalysisResources.TuplesNeedAtLeastTwoElements, nameof(elementNames));
```
Comment (Therzok) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: Message needs to change.

**Comment by TyOverby** [comment]
cc @dotnet/roslyn-compiler

**Comment by TyOverby** [comment]
ping @dotnet/roslyn-compiler

File: src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs [hunk 2]
```diff
@@ -1343,7 +1343,7 @@ internal class CodeAnalysisResources {
         }
         
         /// <summary>
-        ///   Looks up a localized string similar to Windows PDB writer is not available -- could not find Microsoft.DiaSymReader.Native.{0}.dll.
+        ///   Looks up a localized string similar to Windows PDB writer doesn&apos;t support deterministic compilation: &apos;{0}&apos;.
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs]: This doesn't look related to your change. Why is it showing up here?

File: src/Compilers/Core/Portable/CodeAnalysisResources.resx [hunk 3]
```diff
@@ -570,8 +570,8 @@
   <data name="TupleElementLocationCountMismatch" xml:space="preserve">
     <value>If tuple element locations are specified, the number of locations must match the cardinality of the tuple.</value>
   </data>
-  <data name="TuplesNeedAtLeastTwoElements" xml:space="preserve">
-    <value>Tuples must have at least two elements.</value>
+  <data name="TuplesNeedAtLeastOneElement" xml:space="preserve">
+    <value>Tuples must have at least one elements.</value>
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: Element needs to be singular.

File: src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs (hunk 2 above)
Comment (TyOverby) [inline src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs]: My guess is that someone forgot to run the generator, so I wound up with their changes?

File: src/Compilers/Core/Portable/CodeAnalysisResources.resx (hunk 3 above)
Comment (TyOverby) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: Fixed.

**Comment by TyOverby** [comment]
@dotnet-bot test windows_debug_unit32_prtest please

**Comment by TyOverby** [comment]
@dotnet/roslyn-infrastructure previous failure: https://ci.dot.net/job/dotnet_roslyn/job/dev15.6/job/windows_debug_unit32_prtest/396/

**Comment by TyOverby** [comment]
@dotnet/roslyn-infrastructure previous failure: https://ci.dot.net/job/dotnet_roslyn/job/dev15.6/job/windows_coreclr_test_prtest/394/consoleFull#20713614912d31e50d-1517-49fc-92b3-2ca637122019

**Comment by TyOverby** [comment]
@dotnet-bot test windows_coreclr_test_prtest please

**Comment by jcouv** [comment]
FYI @CyrusNajmabadi (in case you want to leverage this from IDE code)

File: src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs]: Looks like it.

**Comment by jaredpar** [comment]
retest windows_debug_unit32_prtest please

**Comment by jaredpar** [comment]
test windows_debug_unit32_prtest please

File: src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs (hunk 2 above)
Comment (tannergooding) [inline src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs]: I think there is a bug with new project system and single file generators. @srivatsn

Comment (srivatsn) [inline src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs]: Ya there's a CPS issue that single file generators doesn't run in all the cases that the old PS did.

File: src/Compilers/Core/Portable/CodeAnalysisResources.resx [hunk 4]
```diff
@@ -570,8 +570,8 @@
   <data name="TupleElementLocationCountMismatch" xml:space="preserve">
     <value>If tuple element locations are specified, the number of locations must match the cardinality of the tuple.</value>
   </data>
-  <data name="TuplesNeedAtLeastTwoElements" xml:space="preserve">
-    <value>Tuples must have at least two elements.</value>
+  <data name="TuplesNeedAtLeastOneElement" xml:space="preserve">
+    <value>Tuples must have at least one element.</value>
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: >Tuples must have at least one element. [](start = 11, length = 38)

The error feels somewhat misleading. Tuples actually cannot have one element. However, tuple type symbol can. Consider using message "Tuple type symbol must have at least one element."

File: src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs [hunk 5]
```diff
@@ -95,8 +95,7 @@ internal sealed class TupleTypeSymbol : WrappedNamedTypeSymbol
             Debug.Assert(elementNames.IsDefault || elementTypes.Length == elementNames.Length);
 
             int numElements = elementTypes.Length;
-
-            if (numElements <= 1)
+            if (numElements == 0)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs]: >if (numElements == 0) [](start = 12, length = 21)

This doesn't feel right, we don't want this method to create types like that.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs]: I am looking at this at the moment.

---
In reply to: [124151349](https://github.com/dotnet/roslyn/pull/20407#discussion_r124151349) [](ancestors = 124151349)

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs]: Ok, I believe this condition should be relaxed only when this helper is called from CommonCreateTupleTypeSymbol. So, either we should create a specialized helper for CommonCreateTupleTypeSymbol to use and share as much code as possible across both of them, or we could add an optional parameter that would explicitly allow the relaxation (the relaxation should be disallowed by default).

File: src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleTypeSymbol.vb [hunk 6]
```diff
@@ -362,7 +362,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Debug.Assert(elementNames.IsDefault OrElse elementTypes.Length = elementNames.Length)
             Dim length As Integer = elementTypes.Length
 
-            If length <= 1 Then
+            If length = 0 Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleTypeSymbol.vb]: >If length = 0 Then [](start = 12, length = 18)

Same comment as for C#.

**Comment by AlekseyTs** [comment]
Done with review pass. I believe the exception message should be adjusted and relaxation in Create method should be explicitly requested rather than done all the time.

File: src/Compilers/Core/Portable/CodeAnalysisResources.resx (hunk 4 above)
Comment (gafter) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: The language does not have a concept of "tuple type symbol", so it would not be appropriate to use that phrase in a language diagnostic.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: > The language does not have a concept of "tuple type symbol", so it would not be appropriate to use that phrase in a language diagnostic.

This is a message for an exception thrown by a public API called **CreateTupleTypeSymbol**, so I do not think it should matter whether language has such concept, and this is not a language diagnostics, it is never produced as a Diagnostics by a compiler.

**Comment by TyOverby** [comment]
@dotnet-bot test windows_debug_unit32_prtest please

File: src/Compilers/Core/Portable/CodeAnalysisResources.resx (hunk 4 above)
Comment (gafter) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: I recommend rewording to describe the error in terms of the argument the ArgumentException is reporting a problem for. E.g. "elementTypes may not be empty".

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: > I recommend rewording to describe the error in terms of the argument the ArgumentException is reporting a problem for. E.g. "elementTypes may not be empty".

I think this would work too.

Comment (TyOverby) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: Text changed to "TupleTypeSymbol elementTypes must not be empty"

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 7]
```diff
@@ -2896,7 +2896,8 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
                 elementNames: elementNames,
                 compilation: this,
                 shouldCheckConstraints: false,
-                errorPositions: default(ImmutableArray<bool>));
+                errorPositions: default(ImmutableArray<bool>),
+                allowOneElementTuple: true);
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: tests for these changes?

File: src/Compilers/Core/Portable/CodeAnalysisResources.resx (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: > Text changed to "TupleTypeSymbol elementTypes must not be empty"

I do not see this change in the PR at the moment.

Comment (TyOverby) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: VS resource designer killed it somehow.  I've changed it again.

File: src/Compilers/Core/Portable/CodeAnalysisResources.resx [hunk 8]
```diff
@@ -571,7 +571,7 @@
     <value>If tuple element locations are specified, the number of locations must match the cardinality of the tuple.</value>
   </data>
   <data name="TuplesNeedAtLeastOneElement" xml:space="preserve">
-    <value>Tuples must have at least one element.</value>
+    <value>TupleTypeSymbol elementTypes must not be empty.</value>
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: Does ```TupleTypeSymbol``` refer to a name of any public type? Otherwise, consider changing to ```Tuple type symbol```, or simply removing the word from the message - "The elementTypes must not be empty."

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 9]
```diff
@@ -920,9 +920,9 @@ public INamedTypeSymbol GetTypeByMetadataName(string fullyQualifiedMetadataName)
                 throw new ArgumentNullException(nameof(elementTypes));
             }
 
-            if (elementTypes.Length <= 1)
+            if (elementTypes.Length == 0)
             {
-                throw new ArgumentException(CodeAnalysisResources.TuplesNeedAtLeastTwoElements, nameof(elementNames));
+                throw new ArgumentException(CodeAnalysisResources.TuplesNeedAtLeastOneElement, nameof(elementNames));
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: The new message complains about ```elementTypes``` and the ```if``` is checking ```elementTypes```, but the exception uses  ```nameof(elementNames)```.

**Comment by TyOverby** [comment]
@AlekseyTs: Where do you want it to be impossible to create 1-element tuples?  I'm doing a review of all the code that calls these methods, and I can't find any place where it makes sense to ban them.

**Comment by jcouv** [comment]
@TyOverby Can the PR move forward again, since Aleksey is back?

**Comment by TyOverby** [comment]
@jcouv Ah yes, thanks for reminding me!

ping @AlekseyTs for my earlier question.

**Comment by AlekseyTs** [comment]
> Where do you want it to be impossible to create 1-element tuples? I'm doing a review of all the code that calls these methods, and I can't find any place where it makes sense to ban them.

I believe none of the existing call sites except the one originated by CreateTupleType API should create 1-element tuples. We only want to change behavior of CreateTupleType API, the rest of the system should work as it was before.

**Comment by jcouv** [comment]
Given that there was no activity on this PR in a while, I'm marking it as "for personal review" for now (so it doesn't show up in our daily list of PRs to review). Thanks

**Comment by gafter** [comment]
Just FYI, this is probably obsoleted by https://github.com/dotnet/roslyn/issues/20648 for which we are likely to do away with `TupleTypeSymbol` as a separate internal API.
